### PR TITLE
Clarify the Clocks chapter of Modelica.Clocked User Guide

### PR DESCRIPTION
**Pull request by pierre-haessig**
Dear MSL maintainers,

While updating OpenModelica and switching to MSL 4.0, I had for the first time an occasion to test `Modelica.Clocked`. I came across the [User guide](https://doc.modelica.org/Modelica%204.0.0/Resources/helpDymola/Modelica_Clocked_UsersGuide.html#Modelica.Clocked.UsersGuide) and I found that the chapter on Clocks could be slightly improved in terms of clarity. This is a first proposition.

The be explicit about my background, I'm a Modelica teacher (not so much Modelica researcher) and I'm not at all an expert of synchronous languages, but I got some notions through colleagues and I had read the Otter *et al.* 2012 paper ["A Library for Synchronous Control Systems in Modelica"](https://2012.international.conference.modelica.org/proceedings/html/pdf/ecp1207627_OtterThieleElmqvist.pdf) once or twice a few years ago. So I had heard of the Clock concept, but also had some time to forget the details. The present proposition comes from self-analyzing my "first time reader" experience.

Since it may be difficult to read through the file diff, here is a summary of my proposed changes.

# 1. Definition of Clock

As a first time reader, I stumbled on the first sentence "A Clock type is a base data type [...] that defines when a particular partition consisting of a set of equations is active" (which is similar to the one in the 2012 paper) because it is 1) quite abstract categorization (like "a rocket is an object") and 2) it uses the library-specific notion of "partition" which is not yet introduced. I tried to provide a more gentle definition although there are probably better ones. Also, you may find it a bit too fuzzy, especially my formulation of "specialized Boolean", but this allow integrating a sentence which was otherwise somehow lost at the end of the document (*"As usual in synchronous languages, a clock is represented by a true value when the clock is active"* which I removed). Now the paragraph introduces both Clock and the notion of partition.

>  The **Clock** type is a base data type introduced in Modelica 3.3 (additionally to Real, Boolean…) which can be understood as a specialized Boolean type: Clock variables are either true (also called "active") or false ("inactive"). It is specialized in the sense that each Clock variable generates a **partition** *(perhaps clarify with "model partition" ?)* consisting of a set of equations which gets activated when the clock is active. In fact, every variable and every equation of a Modelica model (starting with Modelica Language Version 3.3) is assigned to a partition:
> - either the global continuous-time partition (no associated clock)
> - or a clocked partition, associated to exactly one clock (see more details below). 

# 2. Adding more structure

I don't know whether this is allowed in MSL html doc, but I've used heading blocks and list items 

# 3. Clarifying the two types of clocked partitions

As a first time reader, I found the definition of "clocked discrete-time" and "clocked discretized continuous-time" partitions quite hard to understand, since there are defined by negative statements (shall not contain). Also, their name was hidden in the paragraph. In my proposition:
- I highlight the two names of the two partition types using list items
- I keep the definition with negative statements (shall not contain), but with some explanations (adding "no operator relating to continuous time" before proceeding with the enumerations `der`...)
- I add examples of operators that each partition type *may* contain, because I think it clarifies the idea

# 4. Solver paragraph moved

I moved the paragraph on the need to define an integrator for "clocked discretized continuous-time" after the definition of this notion.

# Questions

Shall the previous operator be called "previous" (as is done presently) or "pre" (its syntaxical name). Notice that the derivative operator is name "der"?

Shall "clocked discretized continuous-time" partitions be named simply "clocked continuous-time" partitions?

There are a few repetitions of the fact that Clocks and partitions were introduced in Modelica language 3.3. Would it be better to state this once (at the beginning or the end) and then remove all other references to this fact?

### CODE REVIEW

**Comment by pierre-haessig** [comment]
I'm not sure about the status of CI workflows, but I think that some workflows needs an approval. At least the basic syntax error I introduced in the first two commits are now fixed. @MartinOtter, are you the maintainer of `Modelica.Clocked`?

**Comment by pierre-haessig** [comment]
OK, there was still one error from html_documentation_checks: missing `</li>` closing tags.

**Comment by dietmarw** [comment]
`h1-3` is reserved for tool use. So only use `h4` and `h5`

**Comment by pierre-haessig** [comment]
Thanks @dietmarw for pointing me to MSL's [Documentation on Documentation](https://doc.modelica.org/Modelica%204.0.0/Resources/helpOM/Modelica.UsersGuide.Conventions.Documentation.html). I've fixed the `<h3>` issue.

**Comment by pierre-haessig** [comment]
Now all automated checks are passed! Sorry for needing 6 commits for such a small changes, it's my first contribution to MSL. Do you want me to rebase all this in one commit?
Otherwise, I guess we start the "human-based" review ;-)

**Comment by dietmarw** [comment]
@pierre-haessig Don't worry about the small commits. The PRs are getting squash-merged anyway after they have been approved. I've assigned the relevant [library officers](https://github.com/modelica/MAP-LIB_ProjectRules/blob/master/MAP-LIB_ProjectRules.md#annex-3-library-officers) and you'll just need to wait. It can sometimes take some time.

**Comment by dietmarw** [comment]
@pierre-haessig One thing I noticed is that you submitted this PR from your own fork's master branch. This basically means now that your fork's master branch is kind of locked to this PR until it is merged and can be "released" again. For the future, you should always create a new "topic" branch which then can just live it's live independent of what is going on master and can finally be deleted once the PR is merged.

**Comment by pierre-haessig** [comment]
Thanks for the feedback @dietmarw.

Indeed, using my master branch is a good thing to do in a future case. Now for this specific PR, it almost as if I cloned the MSL repo just for this PR, so I'm not too impacted.

**Comment by pierre-haessig** [comment]
Dear @christoff-buerger, do you think my proposed changes for the Modelica.Clocked User Guide is relevant?

**Comment by christoff-buerger** [comment]
@pierre-haessig: I am very sorry for my late answer. I am on parental leave right now until end of January. Also note that I am only the deputy library officer.

First of all: Thanks a lot for your efforts and feedback!
Second: I will look into detail when back from parental leave.

Now the bad "news": In my opinion, this is going the wrong direction. The idea of the Modelica Standard _Library_ is not to define the Modelica language and its features. For that we have the Modelica Standard. And there are a few ongoing issues regarding clocked in the Modelica specification (e.g., renaming clocked discrete-time and discretized continuous-time partitions to just clocked and discretized partitions etc). I think, the whole descriptive text of the synchronous language features should be removed from the MSL; such are given in the Modelica specification. Such are in `Modelica.Clocked` only for historic reasons right now, because the library has for a very long time been a stand alone separate library (not part of MSL) because the synchronous language features took a long adaption time by Modelica tool vendors.

I really don't want to define and explain Modelica synchronous in the MSL. `Modelica.Clocked` should "just" provide common synchronous support models that can be reused as components and examples of their application, both well documented of course. The documentation of its examples should "just" be about the application domain, like in any other MSL subpackage; but it should not redefine the Modelica language.

**Comment by christoff-buerger** [comment]
> Shall the previous operator be called "previous" (as is done presently) or "pre" (its syntaxical name). Notice that the derivative operator is name "der"?

You are on the wrong track here. Modelica has both operators, `pre` and `previous`. `pre` is about the last value right before an event changed it. `previous` is about the value a clocked variable had at the very end of its last tick's evaluation. Both can be used in clocked partitions; they might yield the same (often the case) but not necessarily (if the clocked partition is a discretized continuous-time partition for example we can have event iteration within a single clock tick's evaluation; only at the very first iteration ist must hold that `pre(v) == previous(v)`).

**Comment by TManikantan** [comment]
@christoff-buerger Just to get perspective, what is the plan with this ticket ,are we going to proceed reviewing it or close it.

**Comment by pierre-haessig** [comment]
Thanks @christoff-buerger for the feedback. I understand the need to "protect" users from low-level implementation details such as partitions in an MSL user guide. On the other hand, I tend to disagree on that referencing to Modelica specification, since I find it a poor learning material (to *understand* the concept). In that sense, the Otter et al. 2012 paper "A Library for Synchronous Control Systems in Modelica" is more useful, except that it's now partially out of date.

Since my PR was a one shot attempt to improve certain paragraphs and since I'm not in position to write a better user guide, I propose to close (reject) the PR.

**Comment by TManikantan** [comment]
@MartinOtter with reference to our discussion, would you please discuss with @hubertus65 on how to proceed with this ticket?

**Comment by MartinOtter** [comment]
Dear Pierre, thanks for this pull request. You are right, the Users Guide of Clocks is not good enough and should be improved so that novice users can grasp the essential details and just use the Modelica.Clocks library without the need to understand all the details defined in the specification. 

I read quickly through your suggestions and this goes in the right direction. However, its not possible to just accept the pull request, because your text contains errors. Most important: Clocks are not special Booleans: E.g. Booleans are piecewise-constant variables and have one or more values at all time instants. Clocks have no value, but they are either active or not active at a time instant. [Section 16.2.1](https://specification.modelica.org/maint/3.6/synchronous-language-elements.html#clocks-and-clocked-variables) of the Modelica Specification 3.6 explains the difference between piecewise-constant variables, clocks and clocked variables in a reasonable way with some figures. In order to plot a clock, a tool might represent the status of a clock as "true" when it is active and as "false" when it is not active. It would be better, if tools would have dedicated values for a clock, say display them as "active" and "not active". Note, even if a clock is plotted with a true value, it cannot have two such "values" at an even instant (contrary to piece-wise constant Booleans, that have, say a false and a true value at one time instant when the Boolean changes from false to true).

Unfortunately, I have currently not enough time to improve your text (maybe in some weeks). So, please wait a bit (or someone else could improve it).

**Comment by MartinOtter** [comment]
> Thanks for the feedback @MartinOtter. What about this reformulation of the first paragraph without mentioning Booleans:
> 
> The **Clock** type is a base data type introduced in Modelica 3.3 (additionally to Real, Boolean…) which allows having sets of equations being activated only at discrete instants, that is "when the clock ticks". Clock variables, at any instant, are either "active" (to represent clock ticks) or "inactive" (between clock ticks). The set of equations which gets activated when its associated clock is active is called a **partition**. In fact, every variable and every equation of a Modelica model (starting with Modelica 3.3) is assigned to a partition:
> 
>     * ...
> 
> 
> Also, in this new proposition, the _purpose_ of clocks is stated directly in the first sentence.

Thanks. Yes, this is a good explanation. Please, include it in the pull request

**Comment by TManikantan** [comment]
@pierre-haessig kindly include the suggestion given above in the PR

File: Modelica/Clocked/UsersGuide/Clocks.mo
Comment (HansOlsson) [inline Modelica/Clocked/UsersGuide/Clocks.mo:90]:
```suggestion
There are two types of sub-clock partitions:
```

File: Modelica/Clocked/UsersGuide/Clocks.mo
Comment (HansOlsson) [inline Modelica/Clocked/UsersGuide/Clocks.mo:87]:
```suggestion
or with the global continuous-time partition).

A sub-clock partition is a part of clock partition that use a specific derived clock,
all equations within a sub-clock partition are active at the same time.
```

File: Modelica/Clocked/UsersGuide/Clocks.mo
Comment (HansOlsson) [inline Modelica/Clocked/UsersGuide/Clocks.mo:99]:
```suggestion
    (it can contain calls of the <strong>previous</strong> operator).</li>
```
One could either say that "it typically contain the previous operator" (as it is normal, but not necessary) or "it can contain the previous operator" (as it is allowed); both mixing them doesn't make sense to me and I think "calls of" clarifies it a bit.

File: Modelica/Clocked/UsersGuide/Clocks.mo
Comment (HansOlsson) [inline Modelica/Clocked/UsersGuide/Clocks.mo:121]: Technically this is now only needed if the sub-clock partition has continuous-time equations; https://github.com/modelica/ModelicaSpecification/pull/3090 but likely to esoteric to mention.

**Comment by HansOlsson** [comment]
I added some comments, and as previously explained by @christoff-buerger we are working on improving the text in the specification as well; and some of the things that are unclear are being worked on. Furthermore, as explained at the MAP-LIB meeting we now have:

- Specification chapter 16
- Modelica Reference; ModelicaReference.Synchronous
- Modelica.Clocked.UsersGuide.Clocks

(The Modelica Reference was also discussed in Aachen.) 
One possibility would be to move the general clock parts to ModelicaReference (and split into sub-parts) and just link to them, and also remove things like Modelica.Clocked.UsersGuide.{ReleaseNotes,Requirements} which is a left-over from having it as a separate library. Obviously the specific library documentation should remain.

There would still be duplication - but the idea is then that ModelicaReference is a quick-guide; similarly as there's https://en.cppreference.com/w/cpp in addition to the real C++specifications. (Except that our specification is freely available.)

**Comment by casella** [comment]
@christoff-buerger, @HansOlsson we have three options at this point
- [ ] Accept this PR as it is
- [ ] Amend it until 02 February, before we branch off 4.1.0-alpha.1
- [ ] Postpone it to 4.2.0

Please indicate which option do you recommend. Thanks!

**Review by HansOlsson — APPROVED** [review]
Seems ok - if Christoff agrees.

**Review by christoff-buerger — APPROVED** [review]
As discussed before: We will just accept the improvements for now, but in the long run the general language feature descriptions will be moved to `ModelicaReference`.

**Comment by christoff-buerger** [comment]
@casella : I integrated all code reviews, reviewed myself, updated from master branch and checked in Dymola. It is all ready to be merged. But our `continuous-integration/jenkins/pr-merge` fails; it looks like it can not fetch from `origin`. This blocks the merge.

Shall I _Merge without waiting for requirements to be met (bypass branch protections)_?

**Review by HansOlsson — APPROVED** [review]

**Comment by Harisankar-Allimangalath** [comment]
@HansOlsson how should we be addressing the check failure .

**Comment by beutlich** [comment]
> @casella : I integrated all code reviews, reviewed myself, updated from master branch and checked in Dymola. It is all ready to be merged. But our `continuous-integration/jenkins/pr-merge` fails; it looks like it can not fetch from `origin`. This blocks the merge.
> 
> Shall I _Merge without waiting for requirements to be met (bypass branch protections)_?

I disabled the jenkins merge checks as merge requirement now.

**Comment by christoff-buerger** [comment]
> > @casella : I integrated all code reviews, reviewed myself, updated from master branch and checked in Dymola. It is all ready to be merged. But our `continuous-integration/jenkins/pr-merge` fails; it looks like it can not fetch from `origin`. This blocks the merge.
> > Shall I _Merge without waiting for requirements to be met (bypass branch protections)_?
> 
> I disabled the jenkins merge checks as merge requirement now.

Thank you!

**Pull request merged**
